### PR TITLE
[ROCm] fastSpecializedAtomicAdd for MI300

### PR DESCRIPTION
**Pull request by jeffdaily**
MI300 adds HW support for packed bfloat16 and fp16. Enable via existing fastSpecializedAtomicAdd.

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

### CODE REVIEW

File: aten/src/ATen/native/cuda/KernelUtils.cuh [hunk 1]
```diff
@@ -5,6 +5,78 @@
...
+  old_val.u32 = __hip_atomic_load((unsigned int*)address, __ATOMIC_RELAXED, __HIP_MEMORY_SCOPE_AGENT);
+  do {
+    new_val.h2r = __hadd2(old_val.h2r, value);
+  } while (!__hip_atomic_compare_exchange_strong(
+        (unsigned int*)address, &old_val.u32, new_val.u32,
+        __ATOMIC_RELAXED, __ATOMIC_RELAXED, __HIP_MEMORY_SCOPE_AGENT));
+  return old_val.h2r;
+#endif
+}
+#define ATOMICADD preview_unsafeAtomicAdd
```
Comment (eqy) [inline aten/src/ATen/native/cuda/KernelUtils.cuh:69]: Are some of these conversions already defined in e.g., https://github.com/pytorch/pytorch/blob/eb0fe029337b31bcb3d4b2d1e539895393975d68/torch/utils/hipify/cuda_to_hip_mappings.py#L540

File: aten/src/ATen/native/cuda/KernelUtils.cuh [hunk 2]
```diff
@@ -5,6 +5,78 @@
...
+#define NATIVE_ZERO_BF16 __float2bfloat16(0.0f)
+#define NATIVE_BFLOAT16 __hip_bfloat16
+#define NATIVE_BFLOAT162 __hip_bfloat162
+#else
+#define ATOMICADD atomicAdd
+#define NATIVE_ZERO_BF16 __int2bfloat16_rz(0)
```
Comment (jianyuh) [inline aten/src/ATen/native/cuda/KernelUtils.cuh:75]: Before RoCM 6.2, it doesn't have `__float2bfloat16` support for MI300x ?

File: aten/src/ATen/native/cuda/KernelUtils.cuh [hunk 3]
```diff
@@ -5,6 +5,78 @@
...
 #endif
 
+// ROCm 6.3 is planned to have these functions, but until then here they are.
+#if defined(USE_ROCM) && ROCM_VERSION >= 60201
+#include <hip/hip_bf16.h>
+#include <hip/hip_fp16.h>
+
+__device__ inline __hip_bfloat162 preview_unsafeAtomicAdd(__hip_bfloat162* address, __hip_bfloat162 value) {
+#if (defined(__gfx940__) || defined(__gfx941__) || defined(__gfx942__)) && \
+  __has_builtin(__builtin_amdgcn_flat_atomic_fadd_v2bf16)
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/KernelUtils.cuh:15]: correct.  trying to use this builtin in rocm 6.2.0 revealed a compiler bug that we quickly fixed for rocm 6.2.1.

File: aten/src/ATen/native/cuda/KernelUtils.cuh [hunk 4]
```diff
@@ -62,17 +133,24 @@ __device__ __forceinline__ void fastSpecializedAtomicAdd(
...
 
   } else {
-    atomicAdd(
-        reinterpret_cast<__half*>(tensor) + index, static_cast<__half>(value));
+#ifdef USE_ROCM
+    gpuAtomicAddNoReturn(
+        reinterpret_cast<at::Half*>(tensor) + index,
+        static_cast<at::Half>(value));
+#else
+    atomicAddNoReturn(
```
Comment (jianyuh) [inline aten/src/ATen/native/cuda/KernelUtils.cuh:150]: `atomicAdd(` ?

Internally this reports the compilation error:

```
1 error detected in the compilation of "fbcode/caffe2/aten/src/ATen/native/cuda/AveragePool3d.cu".
fbcode/caffe2/aten/src/ATen/native/cuda/KernelUtils.cuh(150): error: identifier "atomicAddNoReturn" is undefined
          detected during:
            instantiation of "void at::native::fastSpecializedAtomicAdd(scalar_t *, index_t, index_t, scalar_t) [with scalar_t=c10::Half, index_t=int, <unnamed>=(void *)nullptr]" 
(226): here
            instantiation of "void at::native::fastAtomicAdd(scalar_t *, index_t, index_t, scalar_t, __nv_bool) [with scalar_t=c10::Half, index_t=int]" 
fbcode/caffe2/aten/src/ATen/native/cuda/AveragePool3d.cu(267): here
            instantiation of "void at::native::<unnamed>::avg_pool3d_cuda_update_grad_input_atomic<scalar_t,accscalar_t>(at::PackedTensorAccessor64<const scalar_t, 4UL, at::DefaultPtrTraits>, at::PackedTensorAccessor64<scalar_t, 4UL, at::DefaultPtrTraits>, int, int, int, int, int, int, int, int, int, __nv_bool, int, int, int) [with scalar_t=c10::Half, accscalar_t=float]" 
fbcode/caffe2/aten/src/ATen/native/cuda/AveragePool3d.cu(560): here
```

**Comment by Mellonta** [comment]
Hi Jeff, I tested it internally on ROCm 6.2.0 and the performance looks great—thanks! However, I noticed that the code specifies ROCM_VERSION >= 60201. Is this a requirement, or should it also work with 6.2.0?

**Comment by xw285cornell** [comment]
@Mellonta it's very possible that our internal clang compiler is newer than the clang rpm in  6.2.0

**Comment by jeffdaily** [comment]
> Hi Jeff, I tested it internally on ROCm 6.2.0 and the performance looks great—thanks! However, I noticed that the code specifies ROCM_VERSION >= 60201. Is this a requirement, or should it also work with 6.2.0?

When working on this PR I discovered a bug in our ROCm 6.2 compiler.  To better support you, I got the release team to push the fix as a patch in ROCm 6.2.1.  The compilation will succeed on ROCm 6.2, but your results when using index_add will be garbage for bf16 and fp16 types.  That's why I guard it as needing 6.2.1 or newer.

**Review by xw285cornell — APPROVED** [review]
The failed tests doesn't seem relevant

File: aten/src/ATen/native/cuda/KernelUtils.cuh (hunk 4 above)
Comment (jianyuh) [reply aten/src/ATen/native/cuda/KernelUtils.cuh:150]: Still need to resolve this issue: this `atomicAddNoReturn` caused the failure in the internal phabrictor test suites.

**Review by jianyuh — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cuda/KernelUtils.cuh (hunk 4 above)
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/KernelUtils.cuh]: Fixed now.

File: aten/src/ATen/native/cuda/KernelUtils.cuh (hunk 1 above)
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/KernelUtils.cuh:69]: Good catch.  Fixed now to reduce the size of this PR diff.

**Comment by jeffdaily** [comment]
@jianyuh @xw285cornell Build should be fixed now.

File: aten/src/ATen/native/cuda/KernelUtils.cuh (hunk 2 above)
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/KernelUtils.cuh:73]: We have `__float2bfloat16` in rocm 6.1 and 6.0.  We do not have __int2bfloat16_rz.  That's why I had to use this workaround here.

**Review by jianyuh — APPROVED** [review]

**Pull request closed without merging**